src/caffeine/prif_private_s.f90

```diff
@@ -185,24 +185,45 @@ pure function caf_elem_len(a) result(a_elem_len) bind(C)
        integer(c_size_t), target :: a_elem_len
      end function
 
-     pure function caf_as_int(ptr) bind(C)
-       !! intptr_t caf_as_int(void* ptr);
-       import c_ptr, c_intptr_t
-       type(c_ptr), intent(in), value :: ptr
-       integer(c_intptr_t) :: caf_as_int
-     end function
-
-     pure function caf_as_c_ptr(i) bind(C)
-       !! void* caf_as_c_ptr(intptr_t i);
-       import c_ptr, c_intptr_t
-       integer(c_intptr_t), intent(in), value :: i
-       type(c_ptr) :: caf_as_c_ptr
-     end function
-
   end interface
 
 contains
 
+  pure function caf_as_int(ptr)
+    type(c_ptr), intent(in) :: ptr
+    integer(c_intptr_t) :: caf_as_int
+
+    ! the following snippet ensures at compile time that c_ptr and
+    ! c_intptr_t are actually the same size
+    integer, parameter :: caf_int_ptr_check = merge(c_intptr_t, 0, storage_size(ptr) == storage_size(caf_as_int))
+    integer(caf_int_ptr_check), parameter :: unused = 0_caf_int_ptr_check
+
+    caf_as_int = transfer(ptr, caf_as_int)
+  end function
+
+  pure function caf_as_c_ptr(i)
+    integer(c_intptr_t), intent(in) :: i
+    type(c_ptr) :: caf_as_c_ptr
+
+    caf_as_c_ptr = transfer(i, caf_as_c_ptr)
+  end function
+
+  subroutine caf_base_pointer(coarray_handle, image_num, ptr)
+    type(prif_coarray_handle), intent(in) :: coarray_handle
+    integer(c_int), intent(in) :: image_num
+    integer(c_intptr_t), intent(out) :: ptr
+
+    integer(c_int) :: num_img
+
+    call prif_num_images(num_images=num_img)
+    call assert(image_num .ge. 0 .and. image_num .le. num_img, "caf_base_pointer: image_num not within valid range")
+    if (image_num .eq. 0) then
+      ptr = 0
+    else
+      ptr = caf_convert_base_addr(coarray_handle%info%coarray_data, image_num)
+    end if
```

Comment:
I suspect this branch is now irrelevant.

IIRC `prif_base_pointer()` was specified to accept `image_num==0` and return a null pointer specifically to allow the client to pass the (possibly empty) result of `image_index()` directly to `prif_base_pointer()` before checking the result for validity. 

However the client no longer calls `prif_base_pointer()`, and passing `image_num==0` to the communication routines is an error. As such I think this branch should be removed and the assertion above adjusted to exclude zero as a valid input.

```suggestion
    call assert(image_num .gt. 0 .and. image_num .le. num_img, "caf_base_pointer: image_num not within valid range")
    ptr = caf_convert_base_addr(coarray_handle%info%coarray_data, image_num)
```

Reply:
Agreed, I've now done it this way.